Automattic/wp-calypso
# Staging Sites: Unify staging sites selector

* Closes: https://github.com/Automattic/dotcom-forge/issues/2009
* Closes: https://github.com/Automattic/dotcom-forge/issues/1961

## Proposed Changes

* This PR replaces all the usages of [client/state/selectors/is-staging-site.js](https://github.com/Automattic/wp-calypso/blob/0b1392eeb17355c7fe988ed2aa6d0fb70bd60c64/client/state/selectors/is-staging-site.js) introduced in https://github.com/Automattic/wp-calypso/pull/74257 to use [client/state/selectors/is-site-staging.ts](https://github.com/Automattic/wp-calypso/blob/0b1392eeb17355c7fe988ed2aa6d0fb70bd60c64/client/state/selectors/is-site-staging.ts) instead.
* This PR renames `isSiteStaging` selector to `isSiteWpcomStaging` and uses the updated name in the codebase.
* This PR adds unit tests for [client/state/selectors/is-site-staging.ts](https://github.com/Automattic/wp-calypso/blob/0b1392eeb17355c7fe988ed2aa6d0fb70bd60c64/client/state/selectors/is-site-staging.ts) -> now known as `client/state/selectors/is-site-wpcom-staging.ts`

## Testing Instructions

The first file this PR affects is [client/my-sites/controller.js](https://github.com/Automattic/wp-calypso/blob/0b1392eeb17355c7fe988ed2aa6d0fb70bd60c64/client/my-sites/controller.js). `isSiteStaging` function is used there to check if you are on a staging site and are trying to access pages that staging site should have no access to (for example, `Upgrades`). If you are, it should redirect you to `/home/{site}` URL on a staging site.

**To test**:

1. Make sure you have a staging site 
2. Access the staging site and go to `/posts/`
3. Then click on `Upgrades` in the left sidebar.
4. Observe the URL change to `/home/{site}`


https://user-images.githubusercontent.com/25575134/227190264-5d5862d7-16c9-4dff-955e-a994b301e35b.mov

The second file that is affected by the PR is [client/my-sites/customer-home/cards/actions/quick-links/index.jsx](https://github.com/Automattic/wp-calypso/blob/0b1392eeb17355c7fe988ed2aa6d0fb70bd60c64/client/my-sites/customer-home/cards/actions/quick-links/index.jsx). The Quick Links menu on `/home/{site}` should not show `Add email` or `Add domain` option if the site is staging.

**To test**

1. Make sure you have a staging site
2. Access `/home/{site}` on your staging site
3. Check that the **Quick links** menu on the right has on `Add email` and `Add domain` options:

<img width="1542" alt="Screenshot 2023-03-23 at 12 35 58 PM" src="https://user-images.githubusercontent.com/25575134/227191891-4cb3e367-a306-4742-a240-6c675e241d14.png">

Compared to an Atomic site:

<img width="1556" alt="Screenshot 2023-03-23 at 12 36 50 PM" src="https://user-images.githubusercontent.com/25575134/227192088-98dc9d86-fa12-43c1-bf70-f5bdb79dd978.png">

**To test unit tests**:

The deprecated selector  [client/state/selectors/is-staging-site.js](https://github.com/Automattic/wp-calypso/blob/0b1392eeb17355c7fe988ed2aa6d0fb70bd60c64/client/state/selectors/is-staging-site.js) had unit tests added to it. I removed these unit tests and added new ones for [client/state/selectors/is-site-staging.ts](https://github.com/Automattic/wp-calypso/blob/0b1392eeb17355c7fe988ed2aa6d0fb70bd60c64/client/state/selectors/is-site-staging.ts) instead (now will be known as `client/state/selectors/is-site-wpcom-staging.ts`)

1. Pull this branch to your local Calypso environment
2. Run `yarn test-client -- client/state/selectors/test/is-site-wpcom-staging.js`
3. Check if the tests are passing

**To test `isStagingSite` selector renamed**:

Sidenote: [the issue](https://github.com/Automattic/dotcom-forge/issues/1961) indicates to rename the selector to `isWpcomStagingSite`. I changed it to `isSiteWpcomStaging` instead to be consistent with how other selectors are named e.g. `isSiteWPForTeams` or `isSiteWpcomAtomic`.

1. Go through the staging site creation workflow
2. Check if the site is created properly
3. Go to `/settings/general/{site}` on a staging site
4. Check that `Site Tools` are not visible
5. Go to `Upgrades` in the left sidebar
6. Observe the URL change to `/home/{site}`

## Pre-merge Checklist
